---
title: "Methods of Identifying Cardio-metabolic Risk Factors in Obese Individuals"
author: "Emiline Doerr"
date: "`r Sys.Date()`"
output: html_notebook
---

### Abstract
This analysis was done for a project which looked at the reproducibility of 
bioinformatics methods used in obesity and metabolic health studies.

The study reproduced in this analysis was performed by Rovira et al. Authors applied microarray analysis, miRNA target prediction, pathway enrichment analysis, and statistical analysis to investigate metabolic health and obesity.

### Dataset
The dataset describing the samples collected by Rovira et al. is obtained from the CEL tar file provided for GEO Series GSE169290.
- contains 20 CEL files with non-coding RNA profiling data measured by Affymetrix Multispecies miRNA-4 Array.
- 20 serum samples: 10 metabolically healthy obese (MHO), 10 metabolically unhealthy obese (MUO).


### Contents

1. [Dataset]
1. [Analysis]
    1. [Environment Setup]
    1. [Microarray Analysis]
        1. [Differential expression of miRNAs]
        1. [Find validated target genes]
    1. [Pathway enrichment]
    1. [In silico target prediction]
1. [References]

# Analysis
## Environment setup
Set global variables, install packages, download resource files.
```{r setup}
# Set directories globally
knitr::opts_knit$set(root.dir = '.',
                     resources.dir <- "./resources",
                     data.dir <- "./resources/dataset",
                     results.dir <- "./resources/results")
```

```{r installs, eval=FALSE, message=FALSE, warning=TRUE, include=FALSE}
# Skip by default: set include=TRUE to run chunk if needed.

# Install packages
install.packages("BiocManager")
install.packages("devtools")

# Bioconductor library installs
BiocManager::install(c("GEOquery","affy","oligo", "pd.mirna.4.0","multiMiR",
                       "mirbase.db","miRBaseConverter","pathway"))
```

Get data from GEO with GEOquery package
```{r downloads}
library(GEOquery)

# GEO accession for the dataset
geo.acc <- "GSE169290"

# Download resources from GEO database
geo.data <- getGEO(GEO = geo.acc, destdir = data.dir, GSEMatrix = TRUE) 

# Retrieve matrix data and store it in R object, echo summary
(mirna.eset <- geo.data$GSE169290_series_matrix.txt.gz)

# Experiment data
experimentData(mirna.eset)

# Phenotype data
phenoData(mirna.eset)
varLabels(mirna.eset)
pData(mirna.eset)


# Feature data
featureData(mirna.eset)
varMetadata(mirna.eset)

# Data processing info
mirna.eset$data_processing[1]

# Group labels
mirna.eset$`group:ch1`
```

## Microarray Analysis
### Differential expression of miRNAs 
- TODO! Doesnt match up
```{r diffexpr, message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
library(limma)

# Filter out non-mature miRNAs (accession does not start with MIMA)
#mirna.eset <- subset(mirna.eset, grepl("MIMA*", 
#                                       mirna.eset@featureData@data$Accession))
#dim(mirna.eset)

# Create design matrix from phenotype data
design.matrix <- model.matrix(~0 + mirna.eset$`group:ch1`)

# Name groups in design matrix as MHO and MUO
colnames(design.matrix) <- c("MHO","MUO")

# Switch to compare results
contrast.matrix<-makeContrasts(MHO-MUO, levels=design.matrix)

# Fit to linear model, apply empirical bayes
mirna.lmfit <- lmFit(mirna.eset, design.matrix)
mirna.contr.fit <- contrasts.fit(mirna.lmfit, contrast.matrix)
mirna.fit <- eBayes(mirna.contr.fit)

# Extract differentially expressed miRNAs with topTable
n.features <- dim(mirna.eset)['Features']
mirna.diff.expr <- topTable(mirna.fit, 
                            coef=1, 
                            adjust.method = "none", 
                            p.value = 0.05, 
                            n=n.features,
                            sort.by = "p")

# Show number of diff. expressed miRNA
dim(mirna.diff.expr)

# Volcano plot
volcanoplot(mirna.fit)

```

#### Heatmap
```{r heatmap}
library(pheatmap)


# Make expression matrix, make sample names match up to figure
mirna.eset.2 <- mirna.eset[mirna.diff.expr$ID]

mirna.eset.2$title[mirna.eset.2$`group:ch1` == "unhealthy"]<-gsub(
  "CB",
  "MUO",
  mirna.eset.2$title[mirna.eset.2$`group:ch1` == "unhealthy"])

mirna.eset.2$title[mirna.eset.2$`group:ch1` == "healthy"]<-gsub(
  "CB",
  "MHO",
  mirna.eset.2$title[mirna.eset.2$`group:ch1` == "healthy"])

colnames(mirna.eset.2) <- mirna.eset.2$title
expr <- t(exprs(mirna.eset.2))

pheatmap(expr,
         #cutree_rows = 2,
         scale = "column",
         clustering_distance_rows = "correlation",
         clustering_distance_cols = "correlation",
         fontsize_col = 2)

```

### Find validated target genes
- TODO! Doesnt match up

- Query database for validated targets using miRNA ids
- Show top 5 up and down-regulated miRNAs with predicted gene targets


```{r}
has.targets <- subset(mirna.diff.expr, mirna.diff.expr$Target.Genes != '---')
sw.mti <- subset(has.targets, startsWith(has.targets$Target.Genes,"MTI"))
up <- subset(sw.mti, sw.mti$logFC>0)
down <- subset(sw.mti, sw.mti$logFC<0)

up.5 <- up[1:5,]
# Parse target symbols out of the Target Genes value in ExpressionSet
# Values look like: "word // word // target symbol /// word // word // target symbol /// ...etc
parse_target_data <- function(diff.expr.results) {
  
  i <- 0
  target.lists <- list()
  for (target.genes.str in up.5$Target.Genes) {
    i <- i+1
    target.list <- c()
    for (target in strsplit(target.genes.str, " /// ")) {
      for (info in strsplit(target, " // ")) {
        symbol <- info[3]
        target.list <- append(target.list, target)
      }
    }
    target.lists[[i]] <- unlist(target.list)
  }
  diff.expr.results$Target.Genes <- target.lists
  return(diff.expr.results)
}

up.5 <- parse_target_data(up.5)


```

```{r upreg}
library(multiMiR)

# Differentially expressed miRNAs with positive log2 fold change
mirna.up <- subset(mirna.diff.expr, mirna.diff.expr$logFC>0)
mirbase.query.up <- get_multimir(org = 'hsa', mirna = mirna.up$miRNA_ID, 
                                table = 'mirtarbase', 
                                predicted.cutoff.type = "p")
mirbase.data.up <- mirbase.query.up@data

# Map miRNA ids to results
index <- match(mirna.up$miRNA_ID, mirbase.data.up$mature_mirna_id)
upreg.targets <- mirbase.data.up[index,]

# Generate table with results
mirna.upreg <- as.data.frame(
  cbind(upreg.targets$mature_mirna_id,
        upreg.targets$target_symbol,
        mirna.up$logFC,
        mirna.up$P.Value))
colnames(mirna.upreg) <- c("Name","Target Gene","Log Fold Change","p-value")

# Show top 5
mirna.upreg[1:5,]
```

```{r downreg}
library(multiMiR)
# Differentially expressed miRNA with negative log2 fold change
mirna.down <- subset(mirna.diff.expr, mirna.diff.expr$logFC<0)
mirbase.query.down <- get_multimir(org = 'hsa', mirna = mirna.down$miRNA_ID, 
                                table = 'mirtarbase', 
                                predicted.cutoff.type = "p")
mirbase.data.down <- mirbase.query.down@data

# Map miRNA ids to results
index <- match(mirna.down$miRNA_ID, mirbase.data.down$mature_mirna_id)
downreg.targets <- mirbase.data.down[index,]

# Generate table with results
mirna.downreg <- as.data.frame(
  cbind(downreg.targets$mature_mirna_id,
        downreg.targets$target_symbol,
        mirna.down$logFC,
        mirna.down$P.Value))
colnames(mirna.downreg) <- c("Name","Target Gene","Log Fold Change","p-value")

# Validate results
#expected <- cbind(mirna.down.query@data$target_symbol[index], mirna.down.query@data$mature_mirna_id[index])
#result <- cbind(mirna.upreg$symbols, rownames(mirna.upreg))
#expected == result

# Print top 5
mirna.downreg[1:5,]
```

### Biological Pathway Assignment
- TODO! Finish generating results
```{r}
#BiocManager::install(c("ontologyIndex", "ontologySimilarity"))
#install.packages("paintmap")
library(ontologyIndex)
library(ontologySimilarity)
library(paintmap)
data(go)
data(gene_GO_terms)
data(GO_IC)

all_upreg.go.terms <- gene_GO_terms[upreg.targets$target_symbol]
all_upreg.go.terms <- all_upreg.go.terms[all_upreg.go.terms%in%gene_GO_terms]

# Top 5
top_upreg.go.terms <- gene_GO_terms[mirna.upreg$`Target Gene`[1:5]]
top_upreg.go.terms <- top_upreg.go.terms[top_upreg.go.terms%in%gene_GO_terms]

# Just do top 5 for now
terms <- top_upreg.go.terms
cc <- go$id[go$name == "cellular_component"]
terms_cc <- 
  lapply(terms, function(x) intersection_with_descendants(go, roots=cc, x)) 
data.frame(check.names=FALSE, `#terms`=sapply(terms, length), 
           `#CC terms`=sapply(terms_cc, length))
sim_matrix <- get_sim_grid(
    ontology=go, 
    information_content=GO_IC,
    term_sets=terms)

# TODO - find the output
```


## In silico target prediction
### miRNA target prediction and KEGG pathway analysis
### Target gene analysis with miRDB
- total of 4389 affected genes for top 10 upregulated miRNAs
- selected target genes with target score > 80
- remove duplicated and triplicates
- KEGG pathway analysis on 1.104 predicted target genes
- predicted gene targets signif. enriched for 53 terms

- TODO! ...a lot
```{r}
# Top 10 upregulated miRNAs
top.upreg.mirnas <- mirna.up[1:10,]

# Search miRDB
silico.targets <- get_multimir(mirna = top.upreg.mirnas$miRNA_ID, 
                               org = "hsa", predicted.cutoff = 80, 
                               table="mirdb")
silico.target.data <- as.data.frame(silico.targets@data)
dim(silico.target.data) # Number of affected genes

# remove duplicates and triplicates
silico.target.data <- subset(silico.target.data, 
                             !duplicated(silico.target.data$target_symbol))
dim(silico.target.data)
silico.target.data <- subset(silico.target.data, 
                             !duplicated(silico.target.data$target_symbol))
dim(silico.target.data)

# TODO - keep going here.
```

### KEGG Pathway analysis - not working
```{r}
library(pathview)
# May or may not need
#options(connectionObserver = NULL)

p.values <- as.numeric(silico.target.data$score)
names(p.values) <- silico.target.data$target_ensembl

p.values <- p.values[!is.na(names(p.values))]

# Fails here
pathview(gene.data = -log10(p.values), 
         gne.idtype = "ENSEMBL",
         species = "hsa", 
         pathway.id = silico.target.data$target_entrez, 
         out.suffix = "kegg_pathway_up")

# assign entrez ids to pvalues
p.values <- as.numeric(mirna.downreg$`p-value`)
names(p.values) <- downreg.targets$target_entrez

# remove entries with no entrez ids
pv.out <- pathview(gene.data = -log10(p.values), 
                   pathway.id = silico.target.data$target_entrez,
                   species = "hsa", 
                   out.suffix = "kegg_pathway")
```

#### Results
- TODO! Fix up results (after they are working probably)

Top 40 biological processes + pathways associated with up- and down-regulated miRNAs in MUO individuals
```{r}
```
Upregulated GO/pathway terms specific for target genes 
```{r}
```

# References
```{r}
# Cite paper
# Rovira-Llopis, S., Díaz-Rúa, R., Grau-Del Valle, C., Iannantuoni, F., Abad-Jimenez, Z., Bosch-Sierra, N., Panadero-Romero, J., Victor, V. M., Rocha, M., Morillas, C., & Bañuls, C. (2021). Characterization of Differentially Expressed Circulating miRNAs in Metabolically Healthy versus Unhealthy Obesity. Biomedicines, 9(3), 321. https://doi.org/10.3390/biomedicines9030321

# Cite libraries
knitr::write_bib(x=.packages())

```