---
title: "Methods of Identifying Cardio-metabolic Risk Factors in Obese Individuals"
author: "Emiline Doerr"
date: "`r Sys.Date()`"
output: html_notebook
bibliography: references.bib
biblio-style: "apalike"
link-citations: true
---

This analysis was done for a project which looked at the reproducibility of 
bioinformatics methods used in obesity and metabolic health studies.

The study reproduced in this analysis was performed by Rovira et al. Authors applied microarray analysis, miRNA target prediction, pathway enrichment analysis, and statistical analysis to serum miRNAs which they collected from 10 MHO and 10 MUO individuals. 

The data used in this analysis is provided on GEO. Further description can be found under the data import section below.


Contents:
1. [Environment Setup]
    1. [Install Packages]
    1. [Download Files]
1. [Data processing]
1. [Differential gene expression assessment with limma moderated t-statistics]
1. [References]


## Environment setup

### Install Packages
```{r}
# Installs
install.packages("BiocManager")
install.packages("devtools")

# Bioconductor library installs
BiocManager::install(c("GEOquery","affy","oligo", "pd.mirna.4.0","multiMiR"))

#all libraries
library(BiocManager)
library(devtools)
library(GEOquery)
library(limma)
library(affy)
library(oligo)
library(pd.mirna.4.0)
library(Biobase)
library(preprocessCore)
library(multiMiR)
```

Download the raw CAL tar file from the GEO database.

The dataset describing the samples collected by Rovira et al. is obtained from the CEL tar file provided for GEO Series GSE169290.

The dataset contains CEL files for each of 20 serum samples: 10 metabolically healthy obese (MHO), 10 metabolically unhealthy obese (MUO). 

CEL files contain non-coding RNA profiling data for the sample measured by Affymetrix Multispecies miRNA-4 Array.

The sample IDs, classification (healthy/unhealthy), and miRNA IDs are provided in the supplemental information from Rovira et al.  were translated into a csv file to create a design matrix for the samples. 


### Download Files 
```{r}
# set working directory for download
working_dir <- dirname(rstudioapi::getActiveDocumentContext()$path)
setwd(working_dir)

# Download CEL tar file from GEO database
library(GEOquery)
getGEOSuppFiles("GSE169290")

# Directory with tar
cel_dir <- paste(working_dir, "/GSE169290/", sep = "")
tarfile <- paste(cel_dir, "GSE169290_RAW.tar", sep = "")
untar(tarfile, exdir = cel_dir)

# Untar CEL files to a directory
#untar(paste(raw_dir, "GSE169290_RAW.tar", sep = ""))
#untar("GSE169290_RAW.tar")
```


## Data processing
1. Load data
1. Correct intensity backgrounds by RMA
1. Normalization by quantiles
```{r}
# Find CEL files in working directory
cel_files <- list.celfiles(cel_dir, listGzipped = TRUE)

# Read csv with sample descriptions, add to phenotype data
pdat <- read.csv("phenoData.csv", header = TRUE, row.names = 1)

# Store row names with sample IDs
samplenames <- rownames(pdat)

# Store groups as annotated data frame
metadata <- data.frame(labelDescription="Sample group", row.names="group")
phenodat <- new("AnnotatedDataFrame", data=pdat, varMetadata=metadata)

# Create expression feature set
featureset <- read.celfiles(cel_files, sampleNames = samplenames, phenoData = phenodat)

# Correct intensity backgrounds by RMA + normalize by quantiles
eset <- rma(featureset, background=TRUE, normalize=TRUE)

# Exclude non-mature miRNAs
# Non-mature accession numbers are 'MI0028437',
# mature miRNA sequences are written thus 'MIMAT0033018'
eset <- subset(eset, startsWith(row.names(eset), "MIMA"))

## Differential gene expression with limma moderated t-statistics
library(limma)

# Create design matrix from phenotype data
design <- model.matrix(~0+factor(eset$group))

# Name groups in design matrix as MHO and MUO
colnames(design) <- c("MHO","MUO")

```


# Results with contrast 1
```{r}

# Create contrast matrix
#contrast.matrix<-makeContrasts(MUO-MHO, levels=design)
contrast.matrix<-makeContrasts(MHO-MUO, levels=design)

fit <- contrasts.fit(lmFit(eset, design), contrasts = contrast.matrix)

# Apply empirical Bayes method
ebayes <- eBayes(fit)

(results <- topTable(ebayes, n=1000, sort.by = "B"))

# Lfc
#(results <- topTable(ebayes, n=50000, sort.by = "B", lfc = 1))

```


# Sort by B
```{r}
# Summarize results
(results <- topTable(ebayes, n=5000, sort.by = "T", p.value = 0.92))
upreg <- subset(results, results$logFC>0)
upreg[1:5,]
```


## Upregulated 
```{r}
# Query database for validated targets using miRNA ids
library(multiMiR)

upreg <- subset(results, results$logFC>0)

# Remove ending in miRNA identifiers to match miRTarBase
rownames(upreg) <- gsub("_st","",rownames(upreg))
up_miRNAs <- rownames(upreg)
up <- get_multimir(org = 'hsa', mirna = up_miRNAs, table = 'mirtarbase', summary = TRUE)
updata <- up@data
index <- match(rownames(upreg), updata$mature_mirna_acc)
symbol <- as.character(updata$target_symbol[index])
mirnaids <- as.character(updata$mature_mirna_id[index])
upresult <- cbind(upreg, symbol, mirnaids)
(upresult <- subset(upresult, !is.na(upresult$symbol)))

#up <- get_multimir(org = 'hsa', mirna = up_miRNAs, table = 'targetscan', summary = TRUE)

```

## Downregulated
```{r}

downreg <- subset(results, results$logFC<0)

# Remove ending in miRNA identifiers to match miRTarBase
rownames(downreg) <- gsub("_st","",rownames(downreg))
down_miRNAs <- rownames(downreg)
down <- get_multimir(org = 'hsa', mirna = down_miRNAs, table = 'mirtarbase', summary = TRUE)
downdata <- down@data
index <- match(rownames(downreg), downdata$mature_mirna_acc)
symbol <- as.character(downdata$target_symbol[index])
downresult <- cbind(downreg, symbol)
(downresult <- subset(downresult, !is.na(downresult$symbol)))

```
Notes on miRNA names

## Playing around
```{r}


norm_expr[,1]

```


chunk of code from resource:
```{r}
#pdata <- pData(eset)
#edata <- as.data.frame(exprs(eset))
#fdata <- fData(eset)
#edata <- log2(edata + 1)
#edata <- edata[rowMeans(edata) > 3, ]
#colramp <- colorRampPalette(c(3,"white", 2))(20)
#plot(density(norm_expr[,1]),col=colramp[1],lwd=3,ylim=c(0,.20))
#for(i in 2:20){lines(density(edata[,i]),lwd=3,col=colramp[i])}
```

## Differential gene expression assessment with limma moderated t-statistics

*TODO: map genes*

*TODO: remove top and bottom five*
`#norm_expr <- normalize.quantiles.robust(expr, n.remove = 5)` maybe?
*TODO: Non-mature human miRNAs excluded from downstream analysis*

```{r}
## Differential gene expression with limma moderated t-statistics
library(limma)

# Create design matrix with healthy vs. unhealthy samples
# convert expression data to a matrix
design <- model.matrix(~0+factor(eset[["group"]] == "unhealthy"))
fit <- lmFit(eset, design)
ebayes <- eBayes(fit)
topTable(ebayes, n=10, sort.by = "B")



#groups <- gsub("unhealthy","1",groups)
#groups <- gsub("healthy","0",groups)
#factor <- as.factor(groups)
#design <- model.matrix(~ 0+factor, data=factor)
#colnames(design)<-c("healthy","unhealthy")

# Create contrast matrix
contrast.matrix<-makeContrasts(unhealthy-healthy, levels=design)

# fit the expression data into a linear model
fit <- lmFit(norm_expr, design)

fit

#fit <- lmFit(eset, design)
fit <- contrasts.fit(fit, contrast.matrix)

# Empirical Bayes to adjust variances, computes t-stat, std. error, logFC, log odd
bayes_fit <- eBayes(fit)

# toptable to summarize results
# Used maximum number of genes
results <- topTable(bayes_fit, adjust="BH", number = 50000, sort.by = "B")


upreg <- subset(results, results$logFC>0)
downreg <- subset(results, results$logFC<0)

# top 5 upregulated miRNA
print(upreg[1:5, ])
# top 5 downregulated miRNA
print(downreg[1:5, ])

# Remove ending in miRNA identifiers to match miRTarBase


```
### Upregulated genes
- Rovira et al got 72 upregulated

```{r}

upreg <- subset(results, results$logFC>0)
print(upreg[1:5])

# Remove ending in miRNA identifiers to match miRTarBase
rownames(upreg) <- gsub("_st","",rownames(upreg))
up_miRNAs <- rownames(upreg)

# Query database for validated targets using miRNA ids
library(multiMiR)

(miRNA <- up_miRNAs[1:5])
up <- get_multimir(org = 'hsa', mirna = miRNA, table = 'mirtarbase', summary = TRUE)
updata <- up@data
upindex <- match(updata$mature_mirna_acc, upreg)

upmirna_result <- na.omit(updata$mature_mirna_acc)

match(updata$mature_mirna_acc, upreg)


merge(rownames(upreg), updata$mature_mirna_id[upindex])


multimr_results_up <- get_multimir(org = 'hsa', mirna = up_miRNAs, table = 'mirtarbase', summary = TRUE)
miRtb_data <- multimr_results_up@data
head(miRtb_data)
table <- cbind2(upreg, miRtb_data)
miRtb_data$target_symbol <- sub('NULL', NA, miRtb_data$target_symbol)

select.list()

miRtb_data[rownames()]

#  lucif <- select(example1, keytype = "type", keys = "validated", 
 #                 columns = columns(example1))
#  lucif[grep("Luciferase", lucif$experiment), ]
```


### Downregulated genes
- Rovira et al got 87 downregulated
```{r}

downreg <- subset(results, results$logFC<0)
downreg[1:5]

```

t-statistic for each gene with p-value
*TODO*
```{r}
# t-statistic for each gene with p-value
```

Top 5 up and downregulated in MUO vs MHO (Table 2)
*TODO*
```{r}
```

Volcano plot
*TODO*
[resource](https://biocorecrg.github.io/CRG_RIntroduction/volcano-plots.html)
```{r}

```

## Target prediction with miRTarBase
*if none found, use TargetScanHuman*
```{r}
# Remove ending in miRNA identifiers to match miRTarBase

head(rownames(results))

rownames(results) <- gsub("_st","",rownames(results))
miRNAs <- rownames(results)

# Query database for validated targets using miRNA ids
library(multiMiR)
multimr_results <- get_multimir(org = 'mmu', mirna = miRNAs, table = 'validated', summary = TRUE)
miRtb_data <- multimr_results@data
head(miRtb_data)
colnames(miRtb_data)
head(results)


# add a column to the results table and save to result_final
# some symbols are labelled NULL, substitute NULL to NA
#symbol <- gsub("NULL", NA, miRtb_data$target_symbol)
index <- match(rownames(results),miRtb_data$mature_mirna_acc)
rows <- miRtb_data$mature_mirna_acc[index]


matches <- miRtb_data.row(index)

match(rownames(results),miRtb_data$mature_mirna_acc)

head(symbol)
result_final<-cbind(symbol, results)
cbind.data.frame(results, symbol)
cbind()

upreg <- subset(result_final, result_final$logFC>0)
up$symbol[1:5]
downreg <- subset(result_final, result_final$logFC<0)
down$symbol[1:5]

# write results to csv files
write.csv(result_final,file="DE_result.csv", quote=F)
write.csv(upreg,file="upreg", quote=F)
write.csv(downreg,file="downreg.csv", quote=F)
```

# References
```{r}
# Cite paper
# Rovira-Llopis, S., Díaz-Rúa, R., Grau-Del Valle, C., Iannantuoni, F., Abad-Jimenez, Z., Bosch-Sierra, N., Panadero-Romero, J., Victor, V. M., Rocha, M., Morillas, C., & Bañuls, C. (2021). Characterization of Differentially Expressed Circulating miRNAs in Metabolically Healthy versus Unhealthy Obesity. Biomedicines, 9(3), 321. https://doi.org/10.3390/biomedicines9030321

# Cite libraries
bib <- c(citation("GEOquery"),
         citation("pd.mirna.4.0"), 
         citation("limma"), 
         citation("preprocessCore"), 
         citation("multiMiR"))

print(bib) 
```


