---
title: "Methods of Identifying Cardio-metabolic Risk Factors in Obese Individuals"
author: "Emiline Doerr"
date: "`r Sys.Date()`"
output: html_notebook
---

This analysis was done for a project which looked at the reproducibility of 
bioinformatics methods used in obesity and metabolic health studies.

The study reproduced in this analysis was performed by Rovira et al. Authors applied microarray analysis, miRNA target prediction, pathway enrichment analysis, and statistical analysis to serum miRNAs which they collected from 10 MHO and 10 MUO individuals. 

The data used in this analysis is provided on GEO. Further description can be found under the data import section below.


Contents:

1. [Analysis]
    1. [Environment Setup]
    1. [Data processing]
    1. [Microarray Analysis]
        1. [Mapping miRBase identifiers]
        1. [Differential miRNA expression analysis]
        1. [Find validated target genes]
    1. [Pathway enrichment]
    1. [In silico target prediction]
1. [References]

# Analysis
## Environment setup
Set global variables, install packages, download resource files.

### Dataset
The dataset describing the samples collected by Rovira et al. is obtained from the CEL tar file provided for GEO Series GSE169290.
- contains 20 CEL files with non-coding RNA profiling data measured by Affymetrix Multispecies miRNA-4 Array.
- 20 serum samples: 10 metabolically healthy obese (MHO), 10 metabolically unhealthy obese (MUO). 

```{r installs, message=FALSE, warning=TRUE, include=FALSE}
# Install packages
install.packages("BiocManager")
install.packages("devtools")

# Bioconductor library installs
BiocManager::install(c("GEOquery","affy","oligo", "pd.mirna.4.0","multiMiR",
                       "mirbase.db","miRBaseConverter","pathway"))
```

```{r downloads, include=FALSE}
### Download Files 
library(GEOquery)

# Working directory
knitr::opts_knit$set(root.dir = rstudioapi::getActiveDocumentContext()$path)

# Directory with resource files
resources.dir <- "./resources"
data.dir <- "./resources/dataset"

# GEO accession for the dataset
geo.acc <- "GSE169290"

# Download resources from GEO database
geo.data <- getGEO(GEO = geo.acc, destdir = data.dir, GSEMatrix = TRUE) 
#Retrieve matrix data and store it in R object
# Already normalized
mirna.eset <- geo.data$GSE169290_series_matrix.txt.gz
```

## Microarray Analysis
### Mapping MiRbase identifiers
```{r mapids}
library(miRBaseConverter)

featureNames(mirna.eset) <- gsub("_st","",featureNames(mirna.eset))
mirna.acc <- #subset(featureNames(mirna.eset), startsWith(mirna.acc, "MIMA"))
  featureNames(mirna.eset)

# Convert Accession to ID
acc2name <- miRNA_AccessionToName(mirna.acc)
#featureNames(mirna.eset) <- acc2name$TargetName

# Accessions that have id and are mature (contain miR)
(mirna.map <- subset(acc2name, !is.na(acc2name$TargetName) 
                    & grepl("hsa.*miR",acc2name$TargetName)))
names.index <- match(mirna.map$Accession, mirna.acc)
filtered.eset <- mirna.eset[names.index]
featureNames(filtered.eset) <- mirna.map$TargetName
```

### Differential miRNA expression analysis
```{r diffexpr}
library(limma)

# Filter out non-mature miRNAs (accession does not start with MIMA)
acc <- mirna.eset@featureData@data$Accession
mirna.eset <- subset(mirna.eset, grepl("MIMA*", acc))


# Create design matrix from phenotype data
design.matrix <- model.matrix(~0+factor(mirna.eset$`group:ch1`))

# Name groups in design matrix as MHO and MUO
colnames(design.matrix) <- c("MHO","MUO")

# Switch to compare results
#contrast.matrix<-makeContrasts(MUO-MHO, levels=design)
contrast.matrix<-makeContrasts(MHO-MUO, levels=design.matrix)


expr <- exprs(mirna.eset)

# Fit to linear model, apply empirical bayes
#mirna.lmfit <- lmFit(mirna.eset, design.matrix)
mirna.lmfit <- lmFit(mirna.eset, design.matrix)
mirna.contr.fit <- contrasts.fit(mirna.lmfit, contrast.matrix)
mirna.fit <- eBayes(mirna.contr.fit)


# Extract differentially expressed miRNAs with topTable
n.features <- dim(mirna.eset)['Features']
mirna.diff.expr <- topTable(mirna.fit, 
                            coef=1, 
                            adjust.method = "none", 
                            p.value = 0.1, 
                            n=n.features,
                            #sort.by = "p",
                            confint=TRUE
                            )

# Show number of miRNA
dim(mirna.diff.expr)
head(mirna.diff.expr)

```

```{r volcanoplot, echo=FALSE, message=TRUE, paged.print=TRUE}
print(volcanoplot(mirna.fit, coef=1))
```

```{r heatmap}
expr <- t(exprs(mirna.eset))
rownames(expr) <- paste(gsub("serum_","",mirna.eset$source_name_ch1),
                        gsub("CB","",mirna.eset$title),
                        sep="")

#install.packages("pheatmap")
library(pheatmap)

annotation <- pData(mirna.eset)["group:ch1"]
rownames(annotation) <- rownames(expr)

pheatmap(expr, scale="row", cluster_rows=TRUE, cutree_rows = 2, 
         annotation_row = annotation)

```

### Find validated target genes
- Query database for validated targets using miRNA ids
- Show top 5 up and down-regulated miRNAs with predicted gene targets
```{r upreg}
library(multiMiR)

# Differentially expressed miRNAs with positive log2 fold change
mirna.up <- subset(mirna.diff.expr, mirna.diff.expr$logFC>0)
mirbase.query.up <- get_multimir(org = 'hsa', mirna = mirna.up$miRNA_ID, 
                                table = 'mirtarbase', 
                                predicted.cutoff.type = "p")
mirbase.data.up <- mirbase.query.up@data

# Map miRNA ids to results
index <- match(mirna.up$miRNA_ID, mirbase.data.up$mature_mirna_id)
upreg.targets <- mirbase.data.up[index,]

# Generate table with results
mirna.upreg <- as.data.frame(
  cbind(upreg.targets$mature_mirna_id,
        upreg.targets$target_symbol,
        mirna.up$logFC,
        mirna.up$P.Value))
colnames(mirna.upreg) <- c("Name","Target Gene","Log Fold Change","p-value")

# Print top 5
mirna.upreg[1:5,]

```

```{r downreg}
library(multiMiR)
# Differentially expressed miRNA with negative log2 fold change
mirna.down <- subset(mirna.diff.expr, mirna.diff.expr$logFC<0)
mirbase.query.down <- get_multimir(org = 'hsa', mirna = mirna.down$miRNA_ID, 
                                table = 'mirtarbase', 
                                predicted.cutoff.type = "p")
mirbase.data.down <- mirbase.query.down@data

# Map miRNA ids to results
index <- match(mirna.down$miRNA_ID, mirbase.data.down$mature_mirna_id)
downreg.targets <- mirbase.data.down[index,]

# Generate table with results
mirna.downreg <- as.data.frame(
  cbind(downreg.targets$mature_mirna_id,
        downreg.targets$target_symbol,
        mirna.down$logFC,
        mirna.down$P.Value))
colnames(mirna.downreg) <- c("Name","Target Gene","Log Fold Change","p-value")

# Validate results
#expected <- cbind(mirna.down.query@data$target_symbol[index], mirna.down.query@data$mature_mirna_id[index])
#result <- cbind(mirna.upreg$symbols, rownames(mirna.upreg))
#expected == result

# Print top 5
mirna.downreg[1:5,]
```

### Biological Pathway Assignment
```{r}
#BiocManager::install(c("ontologyIndex", "ontologySimilarity"))
#install.packages("paintmap")
library(ontologyIndex)
library(ontologySimilarity)
library(paintmap)
data(go)
data(gene_GO_terms)
data(GO_IC)

all_upreg.go.terms <- gene_GO_terms[upreg.targets$target_symbol]
all_upreg.go.terms <- all_upreg.go.terms[all_upreg.go.terms%in%gene_GO_terms]

top_upreg.go.terms <- gene_GO_terms[mirna.upreg$`Target Gene`[1:5]]
top_upreg.go.terms <- top_upreg.go.terms[top_upreg.go.terms%in%gene_GO_terms]

terms <- top_upreg.go.terms
cc <- go$id[go$name == "cellular_component"]
terms_cc <- 
  lapply(terms, function(x) intersection_with_descendants(go, roots=cc, x)) 
data.frame(check.names=FALSE, `#terms`=sapply(terms, length), 
           `#CC terms`=sapply(terms_cc, length))
sim_matrix <- get_sim_grid(
    ontology=go, 
    information_content=GO_IC,
    term_sets=terms)



```


## In silico target prediction
### miRNA target prediction and KEGG pathway analysis
### Target gene analysis with miRDB
- total of 4389 affected genes for top 10 upregulated miRNAs
- selected target genes with target score > 80
- remove duplicated and triplicates
- KEGG pathway analysis on 1.104 predicted target genes
- predicted gene targets signif. enriched for 53 terms
```{r}
# Top 10 upregulated miRNAs
top.upreg.mirnas <- mirna.up[1:10,]

silico.targets <- get_multimir(mirna = top.upreg.mirnas$miRNA_ID, 
                               org = "hsa", predicted.cutoff = 80, 
                               table="mirdb")
silico.target.data <- as.data.frame(silico.targets@data)
dim(silico.target.data) # Number of affected genes
# remove duplicates and triplicates
silico.target.data <- subset(silico.target.data, 
                             !duplicated(silico.target.data$target_symbol))
dim(silico.target.data)
silico.target.data <- subset(silico.target.data, 
                             !duplicated(silico.target.data$target_symbol))
dim(silico.target.data)



```

### KEGG Pathway analysis - not working
```{r}
library(pathview)
silico.target.data$target_entrez
silico.target.data$target_ensembl


options(connectionObserver = NULL)

p.values <- as.numeric(silico.target.data$score)
names(p.values) <- silico.target.data$target_ensembl

p.values <- p.values[!is.na(names(p.values))]



pathview(gene.data = -log10(p.values), 
         gne.idtype = "ENSEMBL",
         species = "hsa", 
         pathway.id = silico.target.data$target_entrez, 
         out.suffix = "kegg_pathway_up")

# assign entrez ids to pvalues
p.values <- as.numeric(mirna.downreg$`p-value`)
names(p.values) <- downreg.targets$target_entrez

# remove entries with no entrez ids
pv.out <- pathview(gene.data = -log10(p.values), 
                   pathway.id = "05200", 
                   species = "hsa", 
                   out.suffix = "kegg_pathway_down")

install.packages("ClustVis")
data(gene.idtype.list)
gene.idtype.list
```

# References
```{r}
# Cite paper
# Rovira-Llopis, S., Díaz-Rúa, R., Grau-Del Valle, C., Iannantuoni, F., Abad-Jimenez, Z., Bosch-Sierra, N., Panadero-Romero, J., Victor, V. M., Rocha, M., Morillas, C., & Bañuls, C. (2021). Characterization of Differentially Expressed Circulating miRNAs in Metabolically Healthy versus Unhealthy Obesity. Biomedicines, 9(3), 321. https://doi.org/10.3390/biomedicines9030321

# Cite libraries
bib <- c(citation("GEOquery"),
         citation("pd.mirna.4.0"), 
         citation("limma"), 
         citation("preprocessCore"), 
         citation("multiMiR"))
print(bib)
```