---
title: "Methods of Identifying Cardio-metabolic Risk Factors in Obese Individuals"
author: "Emiline Doerr"
date: "`r Sys.Date()`"
output: html_notebook
---

### Abstract
This analysis was done for a project which looked at the reproducibility of 
bioinformatics methods used in obesity and metabolic health studies.

The study reproduced in this analysis was performed by Rovira et al. Authors applied microarray analysis, miRNA target prediction, pathway enrichment analysis, and statistical analysis to investigate metabolic health and obesity.

### Dataset
The dataset describing the samples collected by Rovira et al. is obtained from the CEL tar file provided for GEO Series GSE169290.
- contains 20 CEL files with non-coding RNA profiling data measured by Affymetrix Multispecies miRNA-4 Array.
- 20 serum samples: 10 metabolically healthy obese (MHO), 10 metabolically unhealthy obese (MUO).


### Contents

1. [Dataset]
1. [Analysis]
    1. [Environment Setup]
    1. [Microarray Analysis]
        1. [Differential expression of miRNAs]
        1. [Find validated target genes]
    1. [Pathway enrichment]
    1. [In silico target prediction]
1. [References]

# Analysis
## Environment setup
Set global variables, install packages, download resource files.
```{r setup}
# Set directories globally
knitr::opts_knit$set(root.dir = '.',
                     resources.dir <- "./resources",
                     data.dir <- "./resources/dataset",
                     results.dir <- "./resources/results")
```

```{r installs, eval=FALSE, message=FALSE, warning=TRUE, include=FALSE}
# Skip by default: set include=TRUE to run chunk if needed.

# Install packages
install.packages("BiocManager")
install.packages("devtools")

# Bioconductor library installs
BiocManager::install(c("GEOquery","affy","oligo","pd.mirna.4.0","multiMiR",
                       "mirbase.db","pathway","gage"))
```

Get data from GEO with GEOquery package
```{r downloads}
library(GEOquery)

# GEO accession for the dataset
geo.acc <- "GSE169290"

# Download resources from GEO database
geo.data <- getGEO(GEO = geo.acc, destdir = data.dir, GSEMatrix = TRUE) 

# Retrieve matrix data and store it in R object, echo summary
mirna.eset <- geo.data$GSE169290_series_matrix.txt.gz
```
# Summarize dataset
```{r eval=FALSE, include=FALSE}
# Experiment data
experimentData(mirna.eset)

# Phenotype data
phenoData(mirna.eset)
varLabels(mirna.eset)
pData(mirna.eset)

# Feature data
featureData(mirna.eset)
varMetadata(mirna.eset)

# Data processing info
mirna.eset$data_processing[1]
mirna.eset$data_row_count

# Group labels
mirna.eset$`group:ch1`
```

## Microarray Analysis
### Differential expression of miRNAs 

```{r diffexpr, echo=TRUE, message=FALSE, warning=FALSE, paged.print=FALSE}
library(limma)

# Filter out non-mature miRNAs (accession does not start with MIMA)
#mirna.eset <- subset(mirna.eset, grepl("MIMA*", 
#                                       mirna.eset@featureData@data$Accession))
#dim(mirna.eset)

# Create design matrix from phenotype data
design.matrix <- model.matrix(~0 + mirna.eset$`group:ch1`)

# Name groups in design matrix as MHO and MUO
colnames(design.matrix) <- c("MHO","MUO")

# Switch to compare results
contrast.matrix<-makeContrasts(MUO-MHO, levels=design.matrix)

# Fit to linear model, apply empirical bayes
mirna.lmfit <- lmFit(mirna.eset, design.matrix)
mirna.contr.fit <- contrasts.fit(mirna.lmfit, contrast.matrix)
mirna.fit <- eBayes(mirna.contr.fit)

# Extract differentially expressed miRNAs with topTable
# p-val < 0.05
n.features <- dim(mirna.eset)['Features']
mirna.diff.expr <- topTable(mirna.fit, 
                            coef=1, 
                            adjust.method = "none", 
                            p.value = 0.05, 
                            n=n.features,
                            sort.by = "p")

# Upregulated and downregulated
mirna.up <- subset(mirna.diff.expr, mirna.diff.expr$logFC>0)
mirna.down <- subset(mirna.diff.expr, mirna.diff.expr$logFC<0)

# Extract differentially expressed miRNA from expression set
mirna.eset.diff.expr <- mirna.eset[mirna.diff.expr$ID]

# Show number of diff. expressed miRNA (comp. to 159)
dim(mirna.diff.expr)

# Volcano plot, label 3 miRNA
volcanoplot(mirna.fit)
```

#### Heatmap
```{r heatmap}
library(pheatmap)

# Extract differentially expressed miRNA from expression set
mirna.eset.2 <- mirna.eset[mirna.diff.expr$ID]

# Modify sample names to show group
mirna.eset.2$title[mirna.eset.2$`group:ch1` == "unhealthy"]<-gsub(
  "CB",
  "MUO",
  mirna.eset.2$title[mirna.eset.2$`group:ch1` == "unhealthy"])

mirna.eset.2$title[mirna.eset.2$`group:ch1` == "healthy"]<-gsub(
  "CB",
  "MHO",
  mirna.eset.2$title[mirna.eset.2$`group:ch1` == "healthy"])
colnames(mirna.eset.2) <- mirna.eset.2$title

# Make expression matrix
expr <- t(exprs(mirna.eset.2))

# Generate heatmap from matrix
# Hierarchical clustering
# Rows centered
# Unit variance scaling applied to rows
# Rows and columns clustered by correlation distance + avg linkage
pheatmap(expr,
         #cutree_rows = 2,
         #scale = "row",
         scale = "column",
         clustering_distance_rows = "correlation",
         clustering_distance_cols = "correlation",
         fontsize_col = 2)

```

### Find validated target genes
- Top 5 up + down successfully replicated

- Query database for validated targets using miRNA ids
- Show top 5 up and down-regulated miRNAs with predicted gene targets


Parsing out targets already included
```{r eval=FALSE, include=FALSE}
#has.targets <- subset(mirna.diff.expr, mirna.diff.expr$Target.Genes != '---')
#sw.mti <- subset(has.targets, startsWith(has.targets$Target.Genes,"MTI"))
#up <- subset(sw.mti, sw.mti$logFC>0)
#down <- subset(sw.mti, sw.mti$logFC<0)

# Differentially expressed miRNAs with positive log2 fold change
upreg <- subset(mirna.diff.expr, mirna.diff.expr$logFC>0)
dim(upreg)

# Differentially expressed miRNAs with negative log2 fold change
downreg <- subset(mirna.diff.expr, mirna.diff.expr$logFC<0)
dim(downreg)

up.5 <- upreg[1:5,]
down.5 <- downreg[1:5,]

# Parse target symbols out of the Target Genes value in ExpressionSet
# Values look like: "word // word // target symbol /// word // word // target symbol /// ...etc
parse_target_data <- function(diff.expr.results) {
  
  i <- 0
  target.lists <- list()
  for (target.genes.str in diff.expr.results$Target.Genes) {
    i <- i+1
    target.list <- c()
    for (target in strsplit(target.genes.str, " /// ")) {
      for (info in strsplit(target, " // ")) {
        symbol <- info[3]
        target.list <- append(target.list, target)
      }
    }
    target.lists[[i]] <- unlist(target.list)
  }
  diff.expr.results$Target.Genes <- target.lists
  return(diff.expr.results)
}

up.5 <- parse_target_data(up.5)
down.5 <- parse_target_data(down.5)

print(up.5[c("miRNA_ID","logFC","P.Value")])
print(down.5[c("miRNA_ID","logFC","P.Value")])

```

- Manually find targets using miRTarBase and TargetScanHuman
Upregulated genes
```{r upreg, eval=FALSE, include=FALSE}
library(multiMiR)

# Query miRTarBase with upregulated miRNA IDs
mirbase.query.up <- get_multimir(org = 'hsa', mirna = mirna.up$miRNA_ID, 
                                table = 'mirtarbase', 
                                predicted.cutoff = 20,
                                predicted.cutoff.type = "p")
mirbase.data.up <- mirbase.query.up@data


# What i tried before... top genes incorrect
# Map miRNA ids to results
index <- match(mirna.up$miRNA_ID, mirbase.data.up$mature_mirna_id)
upreg.targets <- mirbase.data.up[index,]

# Generate table with results
mirna.upreg <- as.data.frame(
  cbind(upreg.targets$mature_mirna_id,
        upreg.targets$target_symbol,
        mirna.up$logFC,
        mirna.up$P.Value))
colnames(mirna.upreg) <- c("Name","Target Gene","Log Fold Change","p-value")

#Print top 5
mirna.upreg[1:5,]

# Extract all target genes for top 5
#up.5 <- mirna.up[1:5,]
#up.5.data <- subset(mirbase.data.up, mirbase.data.up$mature_mirna_id%in%up.5$miRNA_ID)
#uniq.targets.up <- unique(up.5.data$target_symbol)

# Correct target symbols found in the target but not top frequency or otherwise identifiable.

#x.targs <- up.5.data[up.5.data$mature_mirna_id==unique(up.5.data$mature_mirna_id)]
#x.targs <- up.5.data[up.5.data$mature_mirna_id==x,c("target_symbol")]

# Table with frequencies
#freq <- as.data.frame(table(x.targs))

```

Downreg
```{r downreg, eval=FALSE, include=FALSE}
library(multiMiR)
# Differentially expressed miRNA with negative log2 fold change
mirbase.query.down <- get_multimir(org = 'hsa', mirna = mirna.down$miRNA_ID, 
                                table = 'mirtarbase', 
                                predicted.cutoff.type = "p")
mirbase.data.down <- mirbase.query.down@data

# What i tried before... top genes incorrect
# Map miRNA ids to results
index <- match(mirna.down$miRNA_ID, mirbase.data.down$mature_mirna_id)
downreg.targets <- mirbase.data.down[index,]

# Generate table with results
mirna.downreg <- as.data.frame(
  cbind(downreg.targets$mature_mirna_id,
        downreg.targets$target_symbol,
        mirna.down$logFC,
        mirna.down$P.Value))
colnames(mirna.downreg) <- c("Name","Target Gene","Log Fold Change","p-value")

# Validate results
#expected <- cbind(mirna.down.query@data$target_symbol[index], mirna.down.query@data$mature_mirna_id[index])
#result <- cbind(mirna.upreg$symbols, rownames(mirna.upreg))
#expected == result

# Print top 5
mirna.downreg[1:5,]

# Extract all target genes for top 5
#down.5 <- mirna.down[1:5,]
#down.5.data <- subset(mirbase.data.down, mirbase.data.down$mature_mirna_id%in%down.5$miRNA_ID)
#uniq.targets.down <- unique(down.5.data$target_symbol)

# Correct target symbols found in the target but not top frequency or otherwise identifiable.

#x.targs <- down.5.data[down.5.data$mature_mirna_id==unique(down.5.data$mature_mirna_id)]
#x.targs <- down.5.data[down.5.data$mature_mirna_id==x,c("target_symbol")]

# Table with frequencies
#freq <- as.data.frame(table(x.targs))
```

### Biological Pathway Assignment
- Manually with GO, KEGG, Biocarta and WikiPathways databases

[kegg.gs: Common gene set data collections](https://www.rdocumentation.org/packages/gage/versions/2.22.0/topics/kegg.gs)

GO Terms
```{r}
#BiocManager::install(c("ontologyIndex", "ontologySimilarity"))
#install.packages("paintmap")

library(ontologyIndex)
library(ontologySimilarity)
library(paintmap)
data(go)
data(gene_GO_terms)
data(GO_IC)

all_upreg.go.terms <- gene_GO_terms[upreg.targets$target_symbol]
all_upreg.go.terms <- all_upreg.go.terms[all_upreg.go.terms%in%gene_GO_terms]

# Top 5
top_upreg.go.terms <- gene_GO_terms[mirna.upreg$`Target Gene`[1:5]]
top_upreg.go.terms <- top_upreg.go.terms[top_upreg.go.terms%in%gene_GO_terms]

# Just do top 5 for now
terms <- top_upreg.go.terms
cc <- go$id[go$name == "cellular_component"]
terms_cc <- 
  lapply(terms, function(x) intersection_with_descendants(go, roots=cc, x)) 
data.frame(check.names=FALSE, `#terms`=sapply(terms, length), 
           `#CC terms`=sapply(terms_cc, length))
sim_matrix <- get_sim_grid(
    ontology=go, 
    information_content=GO_IC,
    term_sets=terms)


```

All pathway analyses with GAGE
```{r}
BiocManager::install("gage")
library(gage)

# Need expression matrix
# And Entrez symbols for genes

data(kegg.gs)
data(go.gs)
data("carta.gs")  

# ExpressionSet for just top 10 mirna
up.eset <- mirna.eset.2[mirna.eset.2@featureData@data$miRNA_ID%in%up.10$miRNA_ID]
# convert to matrix, set rownames as entrez
up.expr <- exprs(up.eset)
rownames(up.expr) <- mirbase.data.up[1:length(rownames(up.expr)),]$target_symbol
up.expr
gage::sym2eg()

carta.prep <- gagePrep(exprs = up.expr,
              gsets = carta.gs,
                same.dir = TRUE,
              ref = which(startsWith(colnames(up.expr),"MHO")),
              samp = which(startsWith(colnames(up.expr),"MUO")))

carta.up <- gage(exprs = carta.prep,
              gsets = carta.gs,
                same.dir = TRUE,
              #ref = which(startsWith(colnames(carta.prep),"MHO")),
              #samp = which(startsWith(colnames(carta.prep),"MUO"))
              )

carta.up$stats[!is.na(carta.up$stats)]

carta.sum <- gageSum(carta.up)

kegg.up <- gage(exprs = up.expr,
                gsets = kegg.gs,
                same.dir = TRUE,
                ref = which(startsWith(colnames(up.expr),"MHO")),
                samp = which(startsWith(colnames(up.expr),"MUO")))
go.up <- gage(exprs = up.expr,
              gsets = go.gs,
              ref = which(startsWith(colnames(up.expr),"MHO")),
              samp = which(startsWith(colnames(up.expr),"MUO")))

dim(kegg.up$greater)
dim(kegg.up$less)
dim(kegg.up$stats)

carta.up$stats
kegg.up$stats

```

## In silico target prediction
### miRNA target prediction and KEGG pathway analysis
### Target gene analysis with miRDB
```{r}
# Top 10 upregulated miRNAs
up.10 <- mirna.up[1:10,]

# Search miRDB
silico.targets <- get_multimir(mirna = up.10$miRNA_ID, 
                               org = "hsa", 
                               #predicted.cutoff = 80, 
                               table="mirdb")
silico.target.data <- as.data.frame(silico.targets@data)
dim(silico.target.data) # Number of affected genes

# remove duplicates and triplicates
silico.target.data <- subset(silico.target.data, 
                             !duplicated(silico.target.data$target_symbol))
dim(silico.target.data)
silico.target.data <- subset(silico.target.data, 
                             !duplicated(silico.target.data$target_symbol))
dim(silico.target.data)


```
- total of 4389 affected genes for top 10 upregulated miRNAs
- selected target genes with target score > 80
- remove duplicated and triplicates
- KEGG pathway analysis on 1.104 predicted target genes
- predicted gene targets signif. enriched for 53 terms


### KEGG Pathway analysis - not working
Upregulated GO/pathway terms specific for target genes 

```{r}
library(pathview)
# May or may not need
#options(connectionObserver = NULL)

p.values <- as.numeric(silico.target.data$score)
names(p.values) <- silico.target.data$target_ensembl

p.values <- p.values[!is.na(names(p.values))]

# Fails here
pathview(gene.data = -log10(p.values), 
         gne.idtype = "ENSEMBL",
         species = "hsa", 
         pathway.id = silico.target.data$target_entrez, 
         out.suffix = "kegg_pathway_up")




# assign entrez ids to pvalues
#p.values <- as.numeric(mirna.downreg$`p-value`)
#names(p.values) <- downreg.targets$target_entrez

# remove entries with no entrez ids
#pv.out <- pathview(gene.data = -log10(p.values), 
#                   pathway.id = silico.target.data$target_entrez,
#                   species = "hsa", 
#                   out.suffix = "kegg_pathway")
```

# References
```{r}
# Cite paper
# Rovira-Llopis, S., Díaz-Rúa, R., Grau-Del Valle, C., Iannantuoni, F., Abad-Jimenez, Z., Bosch-Sierra, N., Panadero-Romero, J., Victor, V. M., Rocha, M., Morillas, C., & Bañuls, C. (2021). Characterization of Differentially Expressed Circulating miRNAs in Metabolically Healthy versus Unhealthy Obesity. Biomedicines, 9(3), 321. https://doi.org/10.3390/biomedicines9030321

# Cite libraries
knitr::write_bib(x=.packages())

```