---
title: "Methods of Identifying Cardio-metabolic Risk Factors in Obese Individuals"
author: "Emiline Doerr"
date: "`r Sys.Date()`"
output: html_notebook
bibliography: references.bib
biblio-style: "apalike"
link-citations: true
---

This analysis was done for a project which looked at the reproducibility of 
bioinformatics methods used in obesity and metabolic health studies.

The study reproduced in this analysis was performed by Rovira et al. Authors applied microarray analysis, miRNA target prediction, pathway enrichment analysis, and statistical analysis to serum miRNAs which they collected from 10 MHO and 10 MUO individuals. 

The data used in this analysis is provided on GEO. Further description can be found under the data import section below.


Contents:
1. [Environment Setup]
    1. [Install Packages]
    1. [Download Files]
1. [Data processing]
1. [Differential gene expression assessment with limma moderated t-statistics]
1. [References]


## Environment setup

### Install Packages
```{r}
# Installs
install.packages("BiocManager")
install.packages("devtools")

# Bioconductor library installs
BiocManager::install(c("GEOquery","affy","oligo", "pd.mirna.4.0","multiMiR","mirbase.db","miRBaseConverter","pathview"))

#all libraries
library(BiocManager)
library(devtools)
library(GEOquery)
library(limma)
library(affy)
library(oligo)
library(pd.mirna.4.0)
library(Biobase)
library(preprocessCore)
library(multiMiR)
```

Download the raw CAL tar file from the GEO database.

The dataset describing the samples collected by Rovira et al. is obtained from the CEL tar file provided for GEO Series GSE169290.

The dataset contains CEL files for each of 20 serum samples: 10 metabolically healthy obese (MHO), 10 metabolically unhealthy obese (MUO). 

CEL files contain non-coding RNA profiling data for the sample measured by Affymetrix Multispecies miRNA-4 Array.

The sample IDs, classification (healthy/unhealthy), and miRNA IDs are provided in the supplemental information from Rovira et al.  were translated into a csv file to create a design matrix for the samples. 


### Download Files 
```{r}
# set working directory for download
working_dir <- dirname(rstudioapi::getActiveDocumentContext()$path)
setwd(working_dir)

# Download CEL tar file from GEO database
library(GEOquery)
getGEOSuppFiles("GSE169290")

# Directory with tar
cel_dir <- paste(working_dir, "/GSE169290/", sep = "")
tarfile <- paste(cel_dir, "GSE169290_RAW.tar", sep = "")
untar(tarfile, exdir = cel_dir)

# Untar CEL files to a directory
#untar(paste(raw_dir, "GSE169290_RAW.tar", sep = ""))
#untar("GSE169290_RAW.tar")
```


## Data processing
1. Load data
1. Correct intensity backgrounds by RMA
1. Normalization by quantiles
```{r}
library(oligo)

# Find CEL files in directory
working.dir <- dirname(rstudioapi::getActiveDocumentContext()$path)
cel.dir <- paste(working.dir, "/GSE169290/", sep = "")
cel.files <- list.celfiles(listGzipped = TRUE)

# Read csv with sample descriptions, add to phenotype data
pdat <- read.csv("phenoData.csv", header = TRUE, row.names = 1)

# Store row names with sample IDs
samplenames <- rownames(pdat)

# Store groups as annotated data frame
metadata <- data.frame(labelDescription="Sample group", row.names="group")
phenodat <- new("AnnotatedDataFrame", data=pdat, varMetadata=metadata)

# Create expression feature set
mirna.featureset <- read.celfiles(cel.files, sampleNames = samplenames, phenoData = phenodat)

# Correct intensity backgrounds by RMA + normalize by quantiles
mirna.eset <- rma(mirna.featureset, background=TRUE, normalize=TRUE)

```

## Mapping MiRbase identifiers
```{r}
library(miRBaseConverter)

featureNames(mirna.eset) <- gsub("_st","",featureNames(mirna.eset))
mirna.acc <- #subset(featureNames(mirna.eset), startsWith(mirna.acc, "MIMA"))
  featureNames(mirna.eset)

# Convert Accession to ID
acc2name <- miRNA_AccessionToName(mirna.acc)
#featureNames(mirna.eset) <- acc2name$TargetName

# Accessions that have id and are mature (contain miR)
(mirna.map <- subset(acc2name, !is.na(acc2name$TargetName) 
                    & grepl("hsa.*miR",acc2name$TargetName)))
names.index <- match(mirna.map$Accession, mirna.acc)
mirna.map$Accession == mirna.acc[names.index]
mirna.map$Accession == featureNames(mirna.eset[names.index])
filtered.eset <- mirna.eset[names.index]
featureNames(filtered.eset) <- mirna.map$TargetName
```


## Differential miRNA expression assessment with limma moderated t-statistics
```{r}
library(limma)

# Create design matrix from phenotype data
#design.matrix <- model.matrix(~0+factor(mirna.eset$group))
design.matrix <- model.matrix(~0+factor(filtered.eset$group))


# Name groups in design matrix as MHO and MUO
colnames(design.matrix) <- c("MHO","MUO")

# Switch to compare results
#contrast.matrix<-makeContrasts(MUO-MHO, levels=design)
contrast.matrix<-makeContrasts(MHO-MUO, levels=design.matrix)

# Fit to linear model, apply empirical bayes
#mirna.lmfit <- lmFit(mirna.eset, design.matrix)
mirna.lmfit <- lmFit(filtered.eset, design.matrix)
mirna.contr.fit <- contrasts.fit(mirna.lmfit, contrast.matrix)
mirna.fit <- eBayes(mirna.contr.fit)

dim(mirna.fit)

# Extract differentially expressed miRNAs with topTable
mirna.diff.expr <- topTable(mirna.fit, coef=1, adjust.method = "none", n=mirna.n, p.value = 0.05, sort.by = "p")
dim(mirna.diff.expr)

# Show number of miRNA
head(mirna.diff.expr)

```

### Volcano plot (Figure 1)
```{r}
volcanoplot(mirna.fit, coef=1)
```



## Find validated target genes with miRTarBase
- Query database for validated targets using miRNA ids
- if none found, use TargetScanHuman

### Upregulated genes
```{r}
library(multiMiR)

# Differentially expressed miRNAs with positive log2 fold change
mirna.pos <- subset(mirna.diff.expr, mirna.diff.expr$logFC>0)
mirna.pos.query <- get_multimir(org = 'hsa', mirna = rownames(mirna.pos), table = 'mirtarbase', predicted.cutoff.type = "p")
mirna.pos.data <- mirna.pos.query@data

# Map miRNA ids to results
index <- match(rownames(mirna.pos), mirna.pos.data$mature_mirna_id)
upreg.targets <- mirna.pos.data[index,]

# Generate table with results
mirna.upreg <- as.data.frame(
  cbind(upreg.targets$mature_mirna_id,
        upreg.targets$target_symbol,
        mirna.pos$logFC,
        mirna.pos$P.Value))
colnames(mirna.upreg) <- c("Name","Target Gene","Fold Change","p-value")

# Validate results
#expected <- cbind(mirna.pos.query@data$target_symbol[index], mirna.pos.query@data$mature_mirna_id[index])
#result <- cbind(mirna.upreg$symbols, mirna.upreg$names)
#all(expected == result)

# Print top 5
mirna.upreg[1:5,]

```

### Downregulated genes
```{r}
# Differentially expressed miRNA with negative log2 fold change
mirna.neg <- subset(mirna.diff.expr, mirna.diff.expr$logFC<0)
mirna.neg.query <- get_multimir(org = 'hsa', mirna = rownames(mirna.neg), table = 'mirtarbase', predicted.cutoff.type = "p")
mirna.neg.data <- mirna.neg.query@data

# Map miRNA ids to results
index <- match(rownames(mirna.neg), mirna.neg.data$mature_mirna_id)
downreg.targets <- mirna.neg.data[index,]
#cbind(downreg.targets, mirna.neg)


# Generate table with results
mirna.downreg <- as.data.frame(
  cbind(downreg.targets$mature_mirna_id,
        downreg.targets$target_symbol,
        mirna.neg$logFC,
        mirna.neg$P.Value))
colnames(mirna.downreg) <- c("Name","Target Gene","Fold Change","p-value")

# Validate results
#expected <- cbind(mirna.neg.query@data$target_symbol[index], mirna.neg.query@data$mature_mirna_id[index])
#result <- cbind(mirna.upreg$symbols, rownames(mirna.upreg))
#expected == result

# Print top 5
mirna.downreg[1:5,]

```


## miRNA target prediction and KEGG pathway analysis
- used miRBD database
- target score > 80

## Pathway enrichment
- differential miRNAs in MUO vs MHO manually assigned to biological processes
```{r}
library(pathview)
upreg.targets$target_entrez
downreg.targets$target_entrez


# assign entrez ids to pvalues
p.values <- mirna.downreg$`p-value`
names(p.values) <- downreg.targets$target_entrez

# remove entries with no entrez ids
pv.out <- pathview(gene.data = -log10(p.values), 
                   #pathway.id = "05200", 
                   species = "hsa", 
                   out.suffix = "kegg_pathway_down")

#### up-reg #### 
p.values <- up$P.Value

# assign entrez ids to pvalues
names(p.values) <- upsymbol2eg[,2]

# remove entries with no entrez ids
p.values <- p.values[!is.na(names(p.values))]
pv.out <- pathview(gene.data = -log10(p.values), pathway.id = "05200", species = "hsa", out.suffix = "kegg_pathway_up", res = 300)



```


## Target gene analysis with miRDB
- total of 4389 affected genes for top 10 upregulated miRNAs
- selected target genes with target score > 80
- remove duplicated and triplicates
- KEGG pathway analysis on 1.104 predicted target genes
- predicted gene targets signif. enriched for 53 terms


## Results
```{r}
# t-statistic for each gene with p-value
```


### Heatmap (Figure 2)
```{r}
```

### (Figure 3)

# References
```{r}
# Cite paper
# Rovira-Llopis, S., Díaz-Rúa, R., Grau-Del Valle, C., Iannantuoni, F., Abad-Jimenez, Z., Bosch-Sierra, N., Panadero-Romero, J., Victor, V. M., Rocha, M., Morillas, C., & Bañuls, C. (2021). Characterization of Differentially Expressed Circulating miRNAs in Metabolically Healthy versus Unhealthy Obesity. Biomedicines, 9(3), 321. https://doi.org/10.3390/biomedicines9030321

# Cite libraries
bib <- c(citation("GEOquery"),
         citation("pd.mirna.4.0"), 
         citation("limma"), 
         citation("preprocessCore"), 
         citation("multiMiR"))

print(bib)
```


